arch/arm64/src/imx9/imx9_flexcan.c

```diff
@@ -0,0 +1,2086 @@
+
+  /* Always compare also IDE and RTR bits to mask in RX */
+
+  modifyreg32(priv->base + IMX9_CAN_CTRL2_OFFSET, CAN_CTRL2_RETRY_MASK,
+              CAN_CTRL2_RRS | CAN_CTRL2_EACEN |
+              3 << CAN_CTRL2_RETRY_SHIFT);
+
+  /* Clear MB interrupts */
+
+  putreg32(IFLAG1_TX | IFLAG1_RX, priv->base + IMX9_CAN_IFLAG1_OFFSET);
+
+  /* Enable MB interrupts */
+
+  putreg32(IFLAG1_TX | IFLAG1_RX, priv->base + IMX9_CAN_IMASK1_OFFSET);
+
+  /* Set RX buffers to receive */
+
+  for (i = 0; i < RXMBCOUNT; i++)
+    {
+      mb = flexcan_get_mb(priv, i);
+      mb->cs = (CAN_RXMB_EMPTY << CAN_MB_CS_CODE_SHIFT) | CAN_MB_CS_IDE;
+    }
+
+  /* Exit freeze mode */
+
+  if (!imx9_setfreeze(priv->base, false))
+    {
+      canerr("FLEXCAN: unfreeze fail\n");
+      return ERROR;
+    }
+
+  return OK;
+}
+
+/****************************************************************************
+ * Function: imx9_reset
+ *
+ * Description:
+ *   Reset the flexcan and put it into disabled state
+ *
+ * Input Parameters:
+ *   priv - Reference to the private FLEXCAN driver state structure
+ *
+ * Returned Value:
+ *   None
+ *
+ * Assumptions:
+ *
+ ****************************************************************************/
+
+static void imx9_reset(struct imx9_driver_s *priv)
+{
+  modifyreg32(priv->base + IMX9_CAN_MCR_OFFSET, 0, CAN_MCR_SOFTRST);
+
+  if (!imx9_waitmcr_change(priv->base, CAN_MCR_SOFTRST, false))
+    {
+      canerr("Reset failed");
+      return;
+    }
+
+  /* Enable module */
+
+  if (!imx9_setenable(priv->base, false))
+    {
+      canerr("Disable fail\n");
+      return;
+    }
+}
+
+/****************************************************************************
+ * Function: imx9_canpinmux
+ *
+ * Description:
+ *   Mux the pins used for CAN RX&TX
+ *§
+ * Input Parameters:
+ *   None
+ *
+ * Returned Value:
+ *   None
+ *
+ * Assumptions:
+ *
+ ****************************************************************************/
+
+void imx9_canpinmux()
+{
+#ifdef CONFIG_IMX9_FLEXCAN1
+  imx9_iomux_configure(MUX_FLEXCAN1_TX);
+  imx9_iomux_configure(MUX_FLEXCAN1_RX);
+#endif
+
+#ifdef CONFIG_IMX9_FLEXCAN2
+  imx9_iomux_configure(MUX_FLEXCAN2_TX);
+  imx9_iomux_configure(MUX_FLEXCAN2_RX);
+#endif
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Function: imx9_caninitialize
+ *
+ * Description:
+ *   Initialize the CAN controller and driver
+ *
+ * Input Parameters:
+ *   intf - In the case where there are multiple CAN, this value
+ *          identifies which CAN is to be initialized.
+ *
+ * Returned Value:
+ *   OK on success; Negated errno on failure.
+ *
+ * Assumptions:
+ *
+ ****************************************************************************/
+
+int imx9_caninitialize(int intf)
```

Comment:
add static

Reply:
I'd like to keep that function public, so that if one wants to do "CONFIG_NETDEV_LATEINIT", he can call the function from the board files. That is how it is also done in other platforms.